File: macos/Sources/RelicCore/LegalityChecker.swift
import Foundation

public enum CheckStatus: String, Codable, Sendable {
    case incomplete
    case valid
    case wrongOrder
    case invalid
}

public enum IssueKind: String, Codable, Sendable {
    case duplicate
    case conflict
    case unavailable
    case cursePairing
}

public struct CheckIssue: Codable, Hashable, Sendable, Identifiable {
    public let kind: IssueKind
    public let title: String
    public let detail: String
    public let effectIDs: [Int]

    public var id: String { "\(kind.rawValue)-\(effectIDs.map(String.init).joined(separator: "-"))" }
}

public struct CheckResult: Codable, Sendable {
    public let status: CheckStatus
    public let message: String
    public let orderedAffixes: [Affix]
    public let issues: [CheckIssue]
    public let warnings: [CheckIssue]

    public init(
        status: CheckStatus,
        message: String,
        orderedAffixes: [Affix] = [],
        issues: [CheckIssue] = [],
        warnings: [CheckIssue] = []
    ) {
        self.status = status
        self.message = message
        self.orderedAffixes = orderedAffixes
        self.issues = issues
        self.warnings = warnings
    }
}

public struct LegalityChecker: Sendable {
    public init() {}

    public func check(_ affixes: [Affix], mode: CheckMode) -> CheckResult {
        guard affixes.count == 3 else {
            return CheckResult(status: .incomplete, message: "请选择三个词条")
        }

        var issues: [CheckIssue] = []
        var warnings: [CheckIssue] = []

        let duplicates = Dictionary(grouping: affixes, by: \.effectID).values.filter { $0.count > 1 }
        for group in duplicates {
            issues.append(CheckIssue(
                kind: .duplicate,
                title: "词条重复",
                detail: "同一个效果不能在一件遗物上出现两次：\(group[0].name)",
                effectIDs: group.map(\.effectID)
            ))
        }

        let conflictGroups = Dictionary(
            grouping: affixes.filter { $0.compatibilityID != -1 },
            by: \.compatibilityID
        ).values.filter { $0.count > 1 }

        for group in conflictGroups {
            issues.append(CheckIssue(
                kind: .conflict,
                title: "同一互斥池",
                detail: group.map(\.name).joined(separator: "、") + " 不能同时出现",
                effectIDs: group.map(\.effectID)
            ))
        }

        if mode != .compatibilityOnly && !hasPoolAssignment(
            affixes,
            slotPoolSequences: mode.slotPoolSequences
        ) {
            let eligiblePoolIDs = Set(mode.eligiblePoolIDs)
            let unavailable = affixes.filter { affix in
                Set(affix.poolIDs).isDisjoint(with: eligiblePoolIDs)
            }
            let names = unavailable.isEmpty ? affixes.map(\.name) : unavailable.map(\.name)
            issues.append(CheckIssue(
                kind: .unavailable,
                title: unavailable.isEmpty ? "不符合当前槽池模板" : "不在当前出货池",
                detail: names.joined(separator: "、") + (unavailable.isEmpty
                    ? " 无法分配到任一真实的三词条槽池模板"
                    : " 不属于当前校验口径的非零权重出货池"),
                effectIDs: (unavailable.isEmpty ? affixes : unavailable).map(\.effectID)
            ))
        }

        if mode == .deepPositive {
            let curseBound = affixes.filter(\.requiresCurse)
            let curseNames = curseBound.map(\.name).joined(separator: "、")
            let curseRequirement = curseBound.isEmpty
                ? "其中没有仅 A 池词条；"
                : "其中 \(curseBound.count) 条为仅 A 池词条，至少需要 \(curseBound.count) 个对应诅咒槽；"
            let detail = curseRequirement +
                "当前仅预检三条正面效果，完整深夜遗物仍需结合具体遗物 ID、实际槽池模板与负面词条验证。" +
                (curseNames.isEmpty ? "" : "仅 A 池词条：\(curseNames)")
            warnings.append(CheckIssue(
                kind: .cursePairing,
                title: "深夜模式仅作预检",
                detail: detail,
                effectIDs: curseBound.map(\.effectID)
            ))
        }

        let ordered = canonicalOrder(affixes)
        if !issues.isEmpty {
            return CheckResult(
                status: .invalid,
                message: "该三词条组合不合法",
                orderedAffixes: ordered,
                issues: issues,
                warnings: warnings
            )
        }

        if ordered.map(\.effectID) != affixes.map(\.effectID) {
            return CheckResult(
                status: .wrongOrder,
                message: "组合本身可成立，但词条顺序错误",
                orderedAffixes: ordered,
                warnings: warnings
            )
        }

        return CheckResult(
            status: .valid,
            message: mode == .deepPositive
                ? "正面词条预检通过；不等同于完整深夜遗物合法"
                : "该三词条组合合法，顺序正确",
            orderedAffixes: ordered,
            warnings: warnings
        )
    }

    public func canonicalOrder(_ affixes: [Affix]) -> [Affix] {
        affixes.sorted {
            if $0.sortID == $1.sortID { return $0.effectID < $1.effectID }
            return $0.sortID < $1.sortID
        }
    }

    public func hasPoolAssignment(_ affixes: [Affix], slotPools: [Int]) -> Bool {
        guard affixes.count == slotPools.count else { return false }
        if slotPools.isEmpty { return true }

        return permutations(of: Array(affixes.indices)).contains { permutation in
            zip(permutation, slotPools).allSatisfy { index, poolID in
                affixes[index].poolIDs.contains(poolID)
            }
        }
    }

    public func hasPoolAssignment(_ affixes: [Affix], slotPoolSequences: [[Int]]) -> Bool {
        slotPoolSequences.contains { slotPools in
            hasPoolAssignment(affixes, slotPools: slotPools)
        }
    }

    public func randomCombination(from catalog: [Affix], mode: CheckMode) -> [Affix]? {
        let candidates = catalog.filter { $0.isEligible(for: mode) }
        guard candidates.count >= 3 else { return nil }

        for _ in 0..<6_000 {
            let sample = Array(candidates.shuffled().prefix(3))
            let ordered = canonicalOrder(sample)
            let result = check(ordered, mode: mode)
            if result.status == .valid { return ordered }
        }
        return nil
    }

    private func permutations<T>(of values: [T]) -> [[T]] {
        guard let first = values.first else { return [[]] }
        let tail = Array(values.dropFirst())
        return permutations(of: tail).flatMap { permutation in
            (0...permutation.count).map { index in
                var copy = permutation
                copy.insert(first, at: index)
                return copy
            }
        }
    }
}

File: android/rules/src/main/kotlin/com/nightreign/relicchecker/rules/LegalityChecker.kt
package com.nightreign.relicchecker.rules

import kotlin.random.Random

enum class CheckStatus {
    INCOMPLETE,
    VALID,
    WRONG_ORDER,
    INVALID,
}

enum class IssueKind {
    DUPLICATE,
    CONFLICT,
    UNAVAILABLE,
    CURSE_PAIRING,
}

data class CheckIssue(
    val kind: IssueKind,
    val title: String,
    val detail: String,
    val effectIds: List<Int>,
)

data class CheckResult(
    val status: CheckStatus,
    val message: String,
    val orderedAffixes: List<Affix> = emptyList(),
    val issues: List<CheckIssue> = emptyList(),
    val warnings: List<CheckIssue> = emptyList(),
)

class LegalityChecker {
    fun check(affixes: List<Affix>, mode: CheckMode): CheckResult {
        if (affixes.size != SLOT_COUNT) {
            return CheckResult(
                status = CheckStatus.INCOMPLETE,
                message = "请选择三个词条",
            )
        }
        val warnings = deepWarnings(affixes, mode)

        val issues = mutableListOf<CheckIssue>()

        affixes.groupBy(Affix::effectId)
            .values
            .filter { it.size > 1 }
            .forEach { group ->
                issues += CheckIssue(
                    kind = IssueKind.DUPLICATE,
                    title = "词条重复",
                    detail = "同一个效果不能在一件遗物上出现两次：${group.first().name}",
                    effectIds = group.map(Affix::effectId),
                )
            }

        affixes.filter { it.compatibilityId != -1 }
            .groupBy(Affix::compatibilityId)
            .values
            .filter { it.size > 1 }
            .forEach { group ->
                issues += CheckIssue(
                    kind = IssueKind.CONFLICT,
                    title = "同一互斥池",
                    detail = group.joinToString("、", transform = Affix::name) + " 不能同时出现",
                    effectIds = group.map(Affix::effectId),
                )
            }

        if (
            mode != CheckMode.COMPATIBILITY_ONLY &&
            !hasPoolPatternAssignment(affixes, mode.slotPoolSequences)
        ) {
            val unavailable = affixes.filter { affix ->
                affix.poolIds.none(mode.eligiblePoolIds::contains)
            }
            val affected = unavailable.ifEmpty { affixes }
            val isPatternMismatch = unavailable.isEmpty()
            issues += CheckIssue(
                kind = IssueKind.UNAVAILABLE,
                title = if (isPatternMismatch) "不符合当前槽池模板" else "不在当前出货池",
                detail = affected.joinToString("、", transform = Affix::name) +
                    if (isPatternMismatch) {
                        " 无法分配到任一真实的三词条槽池模板"
                    } else {
                        " 不属于当前校验口径的非零权重出货池"
                    },
                effectIds = affected.map(Affix::effectId),
            )
        }

        val ordered = canonicalOrder(affixes)
        if (issues.isNotEmpty()) {
            return CheckResult(
                status = CheckStatus.INVALID,
                message = "该三词条组合不合法",
                orderedAffixes = ordered,
                issues = issues,
                warnings = warnings,
            )
        }

        if (ordered.map(Affix::effectId) != affixes.map(Affix::effectId)) {
            return CheckResult(
                status = CheckStatus.WRONG_ORDER,
                message = "组合本身可成立，但词条顺序错误",
                orderedAffixes = ordered,
                warnings = warnings,
            )
        }

        return CheckResult(
            status = CheckStatus.VALID,
            message = if (mode == CheckMode.DEEP_POSITIVE) {
                "正面词条预检通过；不等同于完整深夜遗物合法"
            } else {
                "该三词条组合合法，顺序正确"
            },
            orderedAffixes = ordered,
            warnings = warnings,
        )
    }

    fun canonicalOrder(affixes: List<Affix>): List<Affix> =
        affixes.sortedWith(compareBy(Affix::sortId, Affix::effectId))

    fun hasPoolAssignment(affixes: List<Affix>, slotPools: List<Int>): Boolean {
        if (affixes.size != slotPools.size) return false
        if (slotPools.isEmpty()) return true

        val used = BooleanArray(affixes.size)
        fun assign(slotIndex: Int): Boolean {
            if (slotIndex == slotPools.size) return true
            for (affixIndex in affixes.indices) {
                if (!used[affixIndex] && slotPools[slotIndex] in affixes[affixIndex].poolIds) {
                    used[affixIndex] = true
                    if (assign(slotIndex + 1)) return true
                    used[affixIndex] = false
                }
            }
            return false
        }
        return assign(0)
    }

    fun hasPoolPatternAssignment(
        affixes: List<Affix>,
        slotPoolSequences: List<List<Int>>,
    ): Boolean = slotPoolSequences.any { hasPoolAssignment(affixes, it) }

    fun randomCombination(
        catalog: List<Affix>,
        mode: CheckMode,
        random: Random = Random.Default,
    ): List<Affix>? {
        val candidates = catalog.filter { it.isEligible(mode) }
        if (candidates.size < SLOT_COUNT) return null

        repeat(MAX_RANDOM_ATTEMPTS) {
            val sample = candidates.shuffled(random).take(SLOT_COUNT)
            val ordered = canonicalOrder(sample)
            if (check(ordered, mode).status == CheckStatus.VALID) return ordered
        }
        return null
    }

    private fun deepWarnings(affixes: List<Affix>, mode: CheckMode): List<CheckIssue> {
        if (mode != CheckMode.DEEP_POSITIVE) return emptyList()
        val curseBound = affixes.filter(Affix::requiresCurse)
        val curseRequirement = if (curseBound.isEmpty()) {
            "其中没有仅 A 池词条；"
        } else {
            "其中 ${curseBound.size} 条为仅 A 池词条，至少需要 ${curseBound.size} 个对应诅咒槽；"
        }
        val curseNames = curseBound.joinToString("、", transform = Affix::name)
        return listOf(
            CheckIssue(
                kind = IssueKind.CURSE_PAIRING,
                title = "深夜模式仅作预检",
                detail = curseRequirement +
                    "当前仅预检三条正面效果，完整深夜遗物仍需结合具体遗物 ID、实际槽池模板与负面词条验证。" +
                    if (curseNames.isEmpty()) "" else "仅 A 池词条：$curseNames",
                effectIds = curseBound.map(Affix::effectId),
            ),
        )
    }

    private companion object {
        const val SLOT_COUNT = 3
        const val MAX_RANDOM_ATTEMPTS = 6_000
    }
}
